clustering/common/src/main/java/org/jboss/as/clustering/controller/RemoveStepHandler.java

```diff
@@ -89,8 +88,8 @@ protected void performRemove(OperationContext context, ModelNode operation, Mode
                 }
             }
 
-            for (Map.Entry<CapabilityReferenceRecorder, Function<PathAddress, String>> entry : this.descriptor.getResourceCapabilityReferences().entrySet()) {
-                entry.getKey().removeCapabilityRequirements(context, resource, null, entry.getValue().apply(address));
+            for (CapabilityReferenceRecorder recorder : registration.getRequirements()) {
```

Comment:
Why don't you use the default recordCapabilitiesAndRequirements method from core AbstractRemoveStepHandler as it is doing the exact same thing from a capability point of view ?

Reply:
A couple of reasons:
* It is *not* doing the exact same thing. The clustering subsystems include capabilities that are conditionally de/registered from/with the OperationContext (e.g. based on whether a specific attribute is defined).  There is no equivalent support in AbstractRemoveStepHandler.
* The clustering remove handler performs capability deregistration prior to removal of the resource, since resolution of capability names may require information from the resource itself.  AbstractRemoveStepHandler does this in reverse order.

Reply:
The resource is passed as a parameter. i guess it is nicer to have 'only' OperationContext as parameter instead of resource + pathAddress
For the specific deregistrations this could be written into the recordCapabilitiesAndRequirements

Reply:
I just wanted to raise that point so that we reduce the specifics of the clustering subsystem. Let's agree that we disagree :) and accept this PR.

Reply:
Don't get me wrong - I complete agree that reducing this custom logic should be a goal.  However, we first need to enhance wildfly-core to cover the use cases for which this custom logic was created.  If we just remove it, we break the existing behavior.